crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -0,0 +1,226 @@
+                    .ok_or(Error::Overflow)?,
+            ),
+        };
+
+        Fulfillment::new(order, executed, fee).map_err(Into::into)
+    }
+
+    fn protocol_fee(&self, prices: ClearingPrices) -> Result<eth::U256, Error> {
+        // TODO: support multiple fee policies
+        if self.order().fee_policies.len() > 1 {
+            return Err(Error::MultipleFeePolicies);
+        }
+
+        let mut protocol_fee = eth::U256::zero();
+        if let Some(fee_policy) = self.order().fee_policies.first() {
+            match fee_policy {
+                FeePolicy::PriceImprovement {
+                    factor,
+                    max_volume_factor,
+                } => {
+                    let price_improvement_fee = self.price_improvement_fee(prices, *factor)?;
+                    let max_volume_fee = self.volume_fee(prices, *max_volume_factor)?;
+                    // take the smaller of the two
+                    protocol_fee = std::cmp::min(price_improvement_fee, max_volume_fee);
+                }
+                FeePolicy::Volume { factor } => {
+                    protocol_fee = self.volume_fee(prices, *factor)?;
+                }
+            }
+        }
+        Ok(protocol_fee)
+    }
+
+    fn price_improvement_fee(
+        &self,
+        prices: ClearingPrices,
+        factor: f64,
+    ) -> Result<eth::U256, Error> {
+        let sell_amount = self.order().sell.amount.0;
+        let buy_amount = self.order().buy.amount.0;
+        let executed = self.executed().0;
+        let surplus_fee = self
+            .dynamic_fee()
+            .map(|fee| fee.0)
+            .ok_or(Error::ProtocolFeeOnStaticOrder)?;
+        match self.order().side {
+            Side::Buy => {
+                // How much `sell_token` we need to sell to buy `executed` amount of `buy_token`
+                let executed_sell_amount = executed
+                    .checked_mul(prices.buy)
+                    .ok_or(Error::Overflow)?
+                    .checked_div(prices.sell)
+                    .ok_or(Error::DivisionByZero)?;
+                // Sell slightly more `sell_token` to capture the `surplus_fee`
+                let executed_sell_amount_with_fee = executed_sell_amount
+                    .checked_add(surplus_fee)
+                    .ok_or(Error::Overflow)?;
+                // Scale to support partially fillable orders
+                let limit_sell_amount = sell_amount
+                    .checked_mul(executed)
+                    .ok_or(Error::Overflow)?
+                    .checked_div(buy_amount)
+                    .ok_or(Error::DivisionByZero)?;
+                // Remaining surplus after fees
+                // Do not return error if `checked_sub` fails because violated limit prices will
+                // be caught by simulation
+                let surplus = limit_sell_amount
+                    .checked_sub(executed_sell_amount_with_fee)
+                    .unwrap_or(eth::U256::zero());
```

Comment:
If this is used for the surplus based fee we need to reference the quoted price somehow.
Otherwise we'll only compare against the signed limit price which already has the slippage baked in. This is not what we want, right?

Reply:
I see the `PriceImprovement` as a general term - improvement over SOME price, whether it's a limit price or some other artificial price (artificial because `driver` works with limit orders technically and doesn't care whether it has slippage baked in and in which amount).
If it were called `MarketPriceImprovement` then it would be misleading IMO. Having a general term is better since we will implement the improvement over quotes 99.99% so renaming it would lead to renaming the autopilot/driver interface which is not too problematic right now, but it could be in the future.
OTOH, I might be biased since I spent a lot of time thinking in those terms, so if you both prefer to have `SurplusFee` I will update it.

Reply:
I've captured the potential renaming https://github.com/cowprotocol/services/issues/2266
I believe this is ok since chances are higher that we will implement this so it's easier to close the issue than to revert the changes.